filter.go

```diff
@@ -8,60 +8,126 @@ import (
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
+type filterEntry struct {
+	f      *net.IPNet
+	reject bool
+}
+
+// Filters is a structure representing a collection of allow/deny
+// net.IPNet filters, together with the RejectByDefault flag, which
+// represents the default filter policy.
+//
+// Note that the last policy added to the Filters is authoritative.
 type Filters struct {
-	mu      sync.RWMutex
-	filters map[string]*net.IPNet
+	mu              sync.RWMutex
+	RejectByDefault bool
+	filters         []*filterEntry
 }
 
+// NewFilters constructs and returns a new set of net.IPNet filters.
+// By default, the new filter rejects no addresses.
 func NewFilters() *Filters {
 	return &Filters{
-		filters: make(map[string]*net.IPNet),
+		RejectByDefault: false,
+		filters:         make([]*filterEntry, 0),
 	}
 }
 
+// AddDialFilter adds a reject rule to the given Filters.  Hosts
+// matching the given net.IPNet filter will be rejected, unless
+// another rule is added which states that they should be accepted.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
 func (fs *Filters) AddDialFilter(f *net.IPNet) {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-	fs.filters[f.String()] = f
+	fs.filters = append(fs.filters, &filterEntry{f: f, reject: true})
+}
+
+// AddAllowFilter adds an accept rule to the given Filters. Hosts
+// matching the given net.IPNet filter will be accepted, unless
+// another policy is added which states that they should be rejected.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
+func (fs *Filters) AddAllowFilter(f *net.IPNet) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+	fs.filters = append(fs.filters, &filterEntry{f: f, reject: false})
 }
 
+// AddrBlocked parses a ma.Multiaddr, and if it can get a valid netip
+// back applies the Filters, returning true if the given address
+// should be rejected, and false if the given address is allowed.
+//
+// If a parsing error occurs, or no filter matches, the Filters
+// default is returned.
 func (f *Filters) AddrBlocked(a ma.Multiaddr) bool {
 	maddr := ma.Split(a)
 	if len(maddr) == 0 {
-		return false
+		return f.RejectByDefault
 	}
 	netaddr, err := manet.ToNetAddr(maddr[0])
 	if err != nil {
 		// if we cant parse it, its probably not blocked
-		return false
+		return f.RejectByDefault
 	}
 	netip := net.ParseIP(netaddr.String())
 	if netip == nil {
-		return false
+		return f.RejectByDefault
 	}
 
 	f.mu.RLock()
 	defer f.mu.RUnlock()
+
+	var reject bool = f.RejectByDefault
+
 	for _, ft := range f.filters {
-		if ft.Contains(netip) {
-			return true
+		if ft.f.Contains(netip) {
+			reject = ft.reject
 		}
 	}
-	return false
+
+	return reject
 }
 
+// Filters returns the list of net.IPNet structs in the given filter
+// rules.
+//
+// FIXME: This function doesn't really make any sense now that filters
```

Comment:
Sure. The semantics of the two don't really make a ton of sense anymore since the filters are order dependent and taking either list independently looses all the semantics but sure.